Fix #15446 Tables added from other databases are not collapsing in the designer section

Tithugues opened the pull request:
The events to link all buttons to collapse tables were initialized at page loading. But when a new table was added, the event wasn't declared for this one.
With this PR, the event is now also declared on table addition.

Fixes: #15446

----
## Roadmap of bugs
- [x] Add new table from other DB, (wrong structure link + on table list)
- [x] Add table and hover (yellow overlay is expected) 
![image](https://user-images.githubusercontent.com/7784660/63378576-cf2d3a00-c392-11e9-85c0-4ca58a92d226.png)

- [x] Add new table from DB and try to move around the table
- [x] Add 2 new tables and try to toggle the hide button (only the clicked table should be toggled)
- [x] Add new table should appear in table list and the number of tables should be updated (i++)
- [x] Hide/show all does not work for added tables
![image](https://user-images.githubusercontent.com/7784660/63379890-8f1b8680-c395-11e9-9314-1459379fe4f8.png)
- [x] Big/small does not work for added tables
![image](https://user-images.githubusercontent.com/7784660/63379939-ace8eb80-c395-11e9-9ea9-00bed4edfe01.png)

williamdes commented:
@sudo-bot Deploy PR

williamdes commented:
@Tithugues When I collapse `film_category` `city` is collapsed, on demo server linked above

And I can not move around the table on the designer

![image](https://user-images.githubusercontent.com/7784660/63356860-c7f13680-c368-11e9-9225-13b780fd340a.png)

Tithugues commented:
@williamdes , thanks for pointing this.

I didn't like the fact to re-declare the `click` function. And in fact, on tables that are not collapsable anymore, the collapse function is called an even number of times (so it toggles it 2 / 4 / 6 / … times). So in fact, it works… but to much…

Options to fix:
* Disable all events before re-enabling all of them (cons: time consumer)
* Enable events only on new items (cons: may require 2 events managements, one for initialization, another for new items)
* Have a function to enable events on one item and call it during initialization and when an item is added (cons: may be time consumer)

I'll firstly see if there is any way to implement the last option.

Tithugues committed:
Fix #15446 Tables added from other databases are not collapsing in the designer section

The events to link all buttons to collapse tables were initialized at page loading. But when a new table was added, the event wasn't declared for this one.
With this commit, the event is now also declared on table addition.

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

williamdes commented:
@sudo-bot /deploy

Tithugues commented:
@williamdes , this should be better now. Can you please confirm that's also OK for you?

I found a second bug: the added table is not added to the "tables list" (first button of the tools tab on the left).
When this will be fixed, we'll also be able to fix a third bug: on mouseover on the added table, the "yellow" background is not set. (Fixing this last one may be possible without fixing the second bug, but may require to change more things. In addition, the second bug seems more important to me.)

Tithugues commented:
One last thing (that may be related to third bug): to drag&drop an added table, you should click on the table name, not on empty part.

Tithugues commented:
But there is still another issue… When adding 2 tables, clicking on button to collapse second table will collapse first table.
I'm looking at this.

Tithugues commented:
> But there is still another issue… When adding 2 tables, clicking on button to collapse second table will collapse first table.
> I'm looking at this.

I found why and I think this is not linked to the changes done in this PR (but it was hidden because the drag&drop / collapse wasn't working). In fact, the `id` and the `table_name` of the added table are not the good one: it's always the first table of the selected DB.

For instance:
```html
<td id="id_zag_sakila.actor" class="tab_zag nowrap tab_zag_noquery" table_name="sakila.actor" query_set="0">
                    <span class="owner">
                        sakila
                    </span>
                    actor_info
                </td>
```
As you can see, the name is `actor_info`, but in the `id` / `table_name` `actor` is set.

williamdes commented:
@Tithugues I think you can add all the fixes in this PR if you want and so we can review all the found bugs one by one before merging


---
Another bug: add a table from another database and click on structure, you will see the following (the database is not the right one it is the one from witch the designer has been initialized.)
This bug should be patched in another PR, I already tried to fix it and it was complex to do an easy fix.
![image](https://user-images.githubusercontent.com/7784660/63370785-304d1180-c383-11e9-8d27-f6a2ba0a80ea.png)

williamdes commented:
@Tithugues I edited the PR description to add a list of bugs ;)

Tithugues committed:
Fix wrong id on table

When a table is added on the fly, the id was always built based on the first table of the targetted DB, not based on the selected table.

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues commented:
> @Tithugues I edited the PR description to add a list of bugs ;)

OK, let's try all in one PR.

Isn't the first and the last bugs almost the same please? If not, what's the point of the "tables list" in the first one please?

Thank you.

williamdes commented:
The last one is what you reported "I found a second bug: the added table is not added to the "tables list" (first button of the tools tab on the left)."

The first one is the structure buttons not leading to the right database
![image](https://user-images.githubusercontent.com/7784660/63377997-832dc580-c391-11e9-8d61-8c68b4d1ab08.png)

I added "+ table list" to be sure the patch works for the two buttons

williamdes commented:
@Tithugues If you do not want to fix some of the bugs feel free to say it :)

Tithugues committed:
Retrieve DB instead of relying on current one

When adding a table to the designer, it is useful to know the DB of this table to be able to jump to its structure.

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues committed:
Be sure to use local variable

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues commented:
I think that the next step is to display the added table to the "table list".
But I don't like the fact to write the HTML in PHP and in JS…

Options to fix:
1. Duplicate the generation of one line in the "tables list" (ugly, isn't it?)
2. Improve JSON content to add the line to add to the "tables list"
3. Move the list generation from PHP to JS

I initially thought about implementing the second option, but was a bit confused when I saw in the source code: `$response->addHTML($html);` while we are in JSON… So I think I'll firstly try to move the list generation to JS.

Tithugues committed:
Build "tables list" in JS

This will allow to have the list generation in only one place, as the tables can now be added on the fly. So JS also needs to add items to this list.

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues committed:
Add missing overlay on added tables

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues committed:
Toggling global collapse of tables now relies on tables items instead of j_tabs

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues commented:
@williamdes , I believe that all the bugs are fixed. Can you please have a look?

Thank you.

Tithugues committed:
Remove variable that is unused anymore

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

williamdes commented:
@sudo-bot Deploy pr

williamdes commented:
@Tithugues when I load the designer I have the following report:
```json
{
    "pma_version": "4.9.1-dev",
    "browser_name": "CHROME",
    "browser_version": "75.0.3770.145",
    "user_os": "Linux",
    "server_software": "nginx/1.10.3",
    "user_agent_string": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.145 Safari/537.36 Vivaldi/2.6.1566.51",
    "locale": "en",
    "configuration_storage": "disabled",
    "php_version": "7.3.6-1+0~20190531112735.39+stretch~1.gbp6131b7",
    "script_name": "db_designer.php",
    "exception_type": "js",
    "exception": {
        "mode": "stack",
        "name": "TypeError",
        "message": "Cannot read property 'style' of null",
        "stack": [
            {
                "func": "Show_new_page_tables",
                "line": "111",
                "column": "21",
                "context": [
                    "    all_tables.prop('checked', check);",
                    "    for (var tab = 0; tab < all_tables.length; tab++) {",
                    "        var input = all_tables[tab];",
                    "        if (input.value) {",
                    "            var element = document.getElementById(input.value);",
                    "            element.style.top = Get_random(550, 20) + 'px';",
                    "            element.style.left = Get_random(700, 20) + 'px';",
                    "            VisibleTab(input, input.value);",
                    "        }",
                    "    }",
                    "    selected_page = -1;"
                ],
                "uri": "js/designer/page.js?v=4.9.1-dev",
                "scriptname": "js/designer/page.js"
            },
            {
                "func": "?",
                "line": "7",
                "column": "13",
                "context": [
                    "    Load_first_page(db, function (page) {",
                    "        if (page) {",
                    "            Load_HTML_for_page(page.pg_nr);",
                    "            selected_page = page.pg_nr;",
                    "        } else {",
                    "            Show_new_page_tables(true);",
                    "        }",
                    "    });",
                    "}",
                    "",
                    "function Save_to_new_page (db, page_name, table_positions, callback) {"
                ],
                "uri": "js/designer/page.js?v=4.9.1-dev",
                "scriptname": "js/designer/page.js"
            },
            {
                "func": "?",
                "line": "100",
                "column": "9",
                "context": [
                    "                if (firstPage === null) {",
                    "                    firstPage = page;",
                    "                }",
                    "            }",
                    "        }",
                    "        callback(firstPage);",
                    "    });",
                    "}",
                    "",
                    "function Show_new_page_tables (check) {",
                    "    var all_tables = $('#id_scroll_tab').find('td input:checkbox');"
                ],
                "uri": "js/designer/page.js?v=4.9.1-dev",
                "scriptname": "js/designer/page.js"
            },
            {
                "func": "IDBTransaction.transaction.oncomplete",
                "line": "62",
                "column": "13",
                "context": [
                    "        var keyRange = IDBKeyRange.lowerBound(0);",
                    "        var cursorRequest = objStore.openCursor(keyRange);",
                    "        var results = [];",
                    "",
                    "        transaction.oncomplete = function (e) {",
                    "            callback(results);",
                    "        };",
                    "",
                    "        cursorRequest.onsuccess = function (e) {",
                    "            var result = e.target.result;",
                    "            if (Boolean(result) === false) {"
                ],
                "uri": "js/designer/database.js?v=4.9.1-dev",
                "scriptname": "js/designer/database.js"
            }
        ],
        "uri": "db_designer.php?"
    }
}
```

Tithugues commented:
@williamdes , do you have this error on https://phpmyadmin-pr-15448.wdes.eu.org / demo environment / locally please?
Any chance for me to reproduce it?

With my DB, I don't have any issue on page loading.

williamdes commented:
> Any chance for me to reproduce it?

You can debug on https://phpmyadmin-pr-15448.wdes.eu.org with the database "public"
Do you want me to send you the data ?

I did no try locally since on https://phpmyadmin-pr-15448.wdes.eu.org the database "sakila" does not have the issue. I think it is data specific

Tithugues committed:
Fix url encoding of table names

Signed-off-by: Hugues Peccatte <hugues.peccatte@gmail.com>

Tithugues commented:
@williamdes , I found the origin of the bug: the table name was not correctly "url encoded" in JavaScript. This should be OK now. (I achieved to reproduce it locally and I don't have the error any more.)

williamdes commented:
@Tithugues very good work !!
could you block adding a table that already exists please ?
It creates bugs and only one of the x duplicates tables will not be saved to the phpmyadmin storage
![image](https://user-images.githubusercontent.com/7784660/63451360-52f32f00-c444-11e9-9f61-042cf9dd0067.png)

williamdes commented:
@Tithugues Or maybe it is a double bug since tables that are added seem not to be saved to the storage ?

williamdes commented:
Another bug `t_x[public.cl%24_shopping_lists]` should be `t_x[public.cl$_shopping_lists]` when sent to `schema_export.php`

Tithugues commented:
> @Tithugues Or maybe it is a double bug since tables that are added seem not to be saved to the storage ?

@williamdes , what does this mean please? Which storage?

williamdes commented:
@Tithugues I mean the `phpmyadmin` database to store the designer or the [local tables](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) into the browser in some edge case where there is no `phpmyadmin storage` configured.

![image](https://user-images.githubusercontent.com/7784660/63452252-1f190900-c446-11e9-8169-801814f13fa7.png)

Tithugues commented:
> Another bug `t_x[public.cl%24_shopping_lists]` should be `t_x[public.cl$_shopping_lists]` when sent to `schema_export.php`

As this bug has not been introduced by the current PR, I prefer to fix it in another PR: https://github.com/phpmyadmin/phpmyadmin/pull/15451

Tithugues commented:
> @Tithugues I mean the `phpmyadmin` database to store the designer or the [local tables](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) into the browser in some edge case where there is no [`phpmyadmin storage`](https://docs.phpmyadmin.net/en/latest/config.html#cfg_Servers_pmadb) configured.
> 
> ![image](https://user-images.githubusercontent.com/7784660/63452252-1f190900-c446-11e9-8169-801814f13fa7.png)

Sorry, but I really don't know what should be stored where and in which case.
Do you know any version where it would work please? So maybe I can compare the behaviors.

Tithugues commented:
One last thing: moving the "tables list" generation from PHP to JS, I ignored the "tab_pos" and "display_page" variables. For now, I don't see any impact, but maybe because I don't know the purpose of these variables…
If you have any information about that, I would be happy to check if I broke anything.
Thank you.

Tithugues commented:
> @Tithugues very good work !!
> could you block adding a table that already exists please ?
> It creates bugs and only one of the x duplicates tables will not be saved to the phpmyadmin storage
> ![image](https://user-images.githubusercontent.com/7784660/63451360-52f32f00-c444-11e9-9f61-042cf9dd0067.png)

As it doesn't seem to come from this PR, I tried to fix it in another PR: https://github.com/phpmyadmin/phpmyadmin/pull/15452
